swarm/benches/connection_handler.rs

```diff
@@ -0,0 +1,357 @@
+                futures::executor::block_on(async {
+                    while swarm_a.any_beh(|b| !b.finished) || swarm_b.any_beh(|b| !b.finished) {
+                        futures::future::select(swarm_b.next(), swarm_a.next()).await;
+                    }
+                });
+            });
+        });
+    }
+}
+
+impl<T: BigBehaviour + NetworkBehaviour> BigBehaviour for libp2p_swarm::Swarm<T> {
+    fn behaviours(&mut self) -> &mut [PollerBehaviour] {
+        self.behaviour_mut().behaviours()
+    }
+}
+
+fn new_swarm<T: NetworkBehaviour>(beh: T) -> libp2p_swarm::Swarm<T> {
+    let swarm_a_keypair = libp2p_identity::Keypair::generate_ed25519();
+    libp2p_swarm::Swarm::new(
+        MemoryTransport::new()
+            .upgrade(multistream_select::Version::V1)
+            .authenticate(libp2p_plaintext::Config::new(&swarm_a_keypair))
+            .multiplex(libp2p_yamux::Config::default())
+            .boxed(),
+        beh,
+        swarm_a_keypair.public().to_peer_id(),
+        libp2p_swarm::Config::with_tokio_executor().with_idle_connection_timeout(Duration::MAX),
+    )
+}
+
+#[derive(Default)]
+struct PollerBehaviour {
+    spam_count: usize,
+    protocols: &'static [StreamProtocol],
+    restarting: bool,
+    finished: bool,
+    other_peer: Option<PeerId>,
+}
+
+#[derive(Debug)]
+struct FinishedSpamming;
+
+impl NetworkBehaviour for PollerBehaviour {
+    type ConnectionHandler = PollerHandler;
+    type ToSwarm = FinishedSpamming;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _peer: libp2p_identity::PeerId,
+        _local_addr: &libp2p_core::Multiaddr,
+        _remote_addr: &libp2p_core::Multiaddr,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(PollerHandler {
+            spam_count: 0,
+            protocols: self.protocols,
+        })
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _peer: libp2p_identity::PeerId,
+        _addr: &libp2p_core::Multiaddr,
+        _role_override: libp2p_core::Endpoint,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        Ok(PollerHandler {
+            spam_count: self.spam_count,
+            protocols: self.protocols,
+        })
+    }
+
+    fn on_swarm_event(&mut self, _: libp2p_swarm::FromSwarm) {}
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: libp2p_identity::PeerId,
+        _connection_id: libp2p_swarm::ConnectionId,
+        _event: libp2p_swarm::THandlerOutEvent<Self>,
+    ) {
+        self.finished = true;
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p_swarm::ToSwarm<Self::ToSwarm, libp2p_swarm::THandlerInEvent<Self>>>
+    {
+        if self.finished && !self.restarting {
+            self.restarting = true;
```

Comment:
What is `restarting` good for? Can this ever be hit?

Reply:
restarting is there to skip initializing connection each iteration of the benchmark